### ISSUE TITLE
prototype normalize bugs 

### ISSUE THREAD

**Issue by zkmake520** [facebookresearch/swav#14]
Hi,  
   In paper, the pseudo-code shows the prototypes are normalized along the first dimension:
   
        **with torch.no_grad():
           C = normalize(C, dim=0, p=2)**
      
   But in the source code, the prototypes are normalized along the second dimension:
          
          **# normalize the prototypes
          with torch.no_grad():
             w = model.module.prototypes.weight.data.clone()
             w = nn.functional.normalize(w, dim=1, p=2)
             model.module.prototypes.weight.copy_(w)**
         
    Since the column of the prototypes is regarded as one cluster,  the prototype should be normalized along the first dimension( dim = 0) ? 
    Thanks

**Comment by mathildecaron31** [comment]
Hi @zkmake520, thanks for your interest in our work.

In the paper, the prototypes are indeed normalized along the first dimension because the prototypes matrix, `C`, is of dimension DxK (i.e, 128x3000).

On the contrary, in the code, `w` is of dimension KxD (i.e, 3000x128). You can easily check that the normalization is done correctly by printing:
```
print(torch.norm(w, dim=1).shape) # should give 3000
print(torch.norm(w, dim=1) # should give a vector with 1 everywhere
``` 
after line 279 in main_swav.py: https://github.com/facebookresearch/swav/blob/master/main_swav.py#L279

Hope that helps
